src/Saml2/Provider.php

```diff
@@ -0,0 +1,389 @@
+     *
+     * @var Request
+     */
+    protected $request;
+
+    /**
+     * The SAML2 User instance.
+     *
+     * @var User
+     */
+    protected $user;
+
+    /**
+     * The configuration.
+     *
+     * @var array
+     */
+    protected $config;
+
+    const CACHE_KEY = 'socialite_saml2_metadata';
+    const CACHE_KEY_TTL = self::CACHE_KEY.'_ttl';
+
+    public function __construct(Request $request)
+    {
+        parent::__construct($request, '', '', '');
+    }
+
+    public function setConfig(ConfigInterface $config): Provider
+    {
+        $config = $config->get();
+
+        $this->config = $config;
+
+        return $this;
+    }
+
+    public static function additionalConfigKeys(): array
+    {
+        return [
+            'client_id',
+            'client_secret',
+            'redirect',
+            'metadata',
+            'ttl',
+            'acs',
+            'entityid',
+            'certificate',
+            'sp_acs',
+        ];
+    }
+
+    protected function getConfig($key = null, $default = null)
+    {
+        if (!empty($key) && empty($this->config[$key])) {
+            return $default;
+        }
+
+        return $key ? Arr::get($this->config, $key, $default) : $this->config;
+    }
+
+    public function redirect()
+    {
+        $this->request->session()->put('state', $state = $this->getState());
+
+        $identityProviderConsumerService = $this->getIdentityProviderEntityDescriptor()
+            ->getFirstIdpSsoDescriptor()
+            ->getFirstSingleSignOnService(SamlConstants::BINDING_SAML2_HTTP_REDIRECT);
+
+        $authnRequest = new AuthnRequest();
+        $authnRequest
+            ->setID(Helper::generateID())
+            ->setProtocolBinding($this->getAssertionConsumerServiceBinding())
+            ->setRelayState($state)
+            ->setIssueInstant(new DateTime())
+            ->setDestination($identityProviderConsumerService->getLocation())
+            ->setNameIDPolicy((new NameIDPolicy())->setFormat(SamlConstants::NAME_ID_FORMAT_PERSISTENT))
+            ->setIssuer(new Issuer($this->getServiceProviderEntityDescriptor()->getEntityID()));
+
+        $redirectBinding = new HttpRedirectBinding();
+
+        $messageContext = new MessageContext();
+        $messageContext->setMessage($authnRequest);
+
+        return $redirectBinding->send($messageContext);
+    }
+
+    protected function getIdentityProviderEntityDescriptorManually(): EntityDescriptor
+    {
+        $acs = $this->getConfig('acs');
+        $entityId = $this->getConfig('entityid');
+        $certificate = $this->getConfig('certificate');
+
+        if (!$entityId || !$certificate) {
+            throw new MissingConfigException('When using "acs", both "entityid" and "certificate" must be set');
+        }
+
+        if (!Str::startsWith($certificate, '-----BEGIN CERTIFICATE-----')) {
+            $certificate = "-----BEGIN CERTIFICATE-----\r\n".
+                implode("\r\n", str_split($certificate, 80)).
```

Comment:
Could we have a comment on the why this magic is done please? :)

Reply:
Yup, and I actually realised the library can provide this magic so I used the library method and commented as well